### PR TITLE
Python 3.13.  Remove ruff flakiness

### PR DESCRIPTION
**Pull request by fyellin**
Added test for Python 3.13.  (Is it time to remove support for 3.8??)

Codegen was giving me occasional bizarre results because of ruff.  Fixed by using a different file each time for "ruff" rather than reusing the same file over and over.

Finding and fixing the bug above required me to print out codegen deltas in a more easily understandable way.  I'm keeping those changes.

### CODE REVIEW

**Comment by fyellin** [comment]
Python 3.8 reached end-of-live on 8 October, 2024.  I can pull the plug on 3.8 at the same time I add 3.13, if you want.

**Review by almarklein — COMMENTED** [review]
I suspect this has to do with ruff caching. When I looked up the cli arg to disable caching, I noticed that ruff can also run via stdin/stdout, avoiding the need for a temporary file. Here's a minimal working example:

```py
import subprocess
import sys
FormatError = Exception

cmd = [
        sys.executable,
        "-m",
        "ruff",
        "format",
        "--line-length",
        str(88),
        "--no-cache",
        "--stdin-filename",
        "tmp.py"
        
    ]
input = """
foo  = 3         + 1 8
"""

p = subprocess.run(cmd, input=input.encode(), stdout=subprocess.PIPE, stderr=subprocess.PIPE)

if p.returncode:
    stdout = p.stdout.decode(errors="ignore")
    stderr = p.stderr.decode(errors="ignore")
    raise FormatError(
       f"Could not format source ({p.returncode}).\n\nstdout: "
        + stdout
        + "\n\nstderr: "
        + stderr
    )
else:
    result = p.stdout.decode(errors="ignore")

print(result)
``

I also separeted stdout and stderr, as it can be problematic, as seen in https://github.com/pygfx/pygfx/pull/862

**Comment by almarklein** [comment]
>  I can pull the plug on 3.8 at the same time I add 3.13, if you want.

Let's do that too :)

File: codegen/tests/test_codegen_z.py [hunk 1]
```diff
@@ -3,7 +3,9 @@
 """
 
 import os
+import sys
 import time
+from difflib import unified_diff
```
[the remark this answers is not in the document]
Comment (Korijn) [reply codegen/tests/test_codegen_z.py:8]: It's stdlib right? https://docs.python.org/3/library/difflib.html

[the remark this answers is not in the document]
Comment (almarklein) [reply codegen/tests/test_codegen_z.py:8]: Oh, learned something new today. Awesome!

**Comment by fyellin** [comment]
Removed 3.8.  I suppose we can now use | and |= on dictionaries, and str.removeprefix.

**Comment by fyellin** [comment]
I was wrong.  It requires an administrator to remove the 3.8 requirement.  The branch protection rule requiring that the test py38 pass can only be removed by an administrator, and only an administrator can add the requirement for py313. [As a contributor, I can't even see these settings.]

I've undone my 3.8 -> 3.9 changes.  If someone else wants to make them, just update ci.yml, pyproject.toml, and the comment in start.rst

Other changes have been made.  ruff now uses stdin/stdout and seems to be doing the right thing.

**Comment by fyellin** [comment]
This is ready for review and merging.  

It tests both Python 3.8 and 3.13.  It also includes the fixes to get ruff to work through stdin.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -160,7 +160,7 @@ jobs:
         include:
           - name: Test Linux py38
             os: ubuntu-latest
-            pyversion: '3.8'
+            pyversion: '3.9'
```
Comment (almarklein) [inline .github/workflows/ci.yml:163]: Let's match the version with the title - until we remove this soon :)
```suggestion
            pyversion: '3.8'
```

File: codegen/utils.py [hunk 3]
```diff
@@ -123,14 +118,24 @@ def format_code(src, singleline=False):
...
         "--line-length",
         str(line_length),
-        tempfilename,
+        "--no-cache",
+        "--stdin-filename",
+        "tmp.py",
     ]
-    p = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+
+    p = subprocess.run(cmd, input=src.encode(), capture_output=True)
```
Comment (almarklein) [inline codegen/utils.py:126]: Nice, I did not know this arg:

> If capture_output is true, stdout and stderr will be captured. When used, the internal [Popen](https://docs.python.org/3/library/subprocess.html#subprocess.Popen) object is automatically created with stdout and stderr both set to [PIPE](https://docs.python.org/3/library/subprocess.html#subprocess.PIPE).

**Review by almarklein — DISMISSED** [review]
Looking good!

**Review by almarklein — APPROVED** [review]

**Review by Korijn — APPROVED** [review]

**Pull request merged**